tmpl/wiki/package-quality.md

```diff
@@ -0,0 +1,138 @@
+## Package Quality
+
+The MirageOS ecosystem of libraries is booming:
+
+```
+$ opam list -a | grep mirage | wc -l
+      90
+```
+
+We now have various libraries covering all aspects of the OS: from network to
+block device drivers -- at all aspects of the stack: from virtual network cards
+to TLS or virtual block devices to virtual filesystems. These libraries also
+span multiple hypervisors including Xen and (via solo5) KVM.
+
+All of this is great: this means that as a MirageOS user, you can spend time
+developing your own application instead of spending time re-implementing
+something that the OS would usually do for you. There are of course still
+gaps in the ecosystem, but things are getting better!
+
+The next problem is knowing the level of confidence to attribute to each of
+these libraries. Would be great to be able to answer questions such as:
+Does the library has a good test coverage? Was it formally verified?
+Was it used in production? In order to do this, I would like to introduce
+the "MirageOS quality levels". Ideally this will live in a tool (e.g.
+`mirage lint`) but to start we list the criteria that package aiming
+for increased quality should follow.
+
+### ![Level 1:](https://img.shields.io/badge/level-1-blue.svg) Follow Packaging Guidelines
+
+See the [packaging guidelines](https://mirage.io/wiki/packaging) for more
+details. In summary:
+
+- The package should work well with [odig](http://erratique.ch/software/odig)
+  e.g. `CHANGES.md`, `LICENSE.md` and `README.md` should exist.
+
+- The package should work well with [topkg](http://erratique.ch/software/topkg),
+  e.g. `pkg/pkg.ml` should exist.
+
+- The use of [jbuilder](http://jbuilder.readthedocs.io/en/latest/) is
+  strongly encouraged but not yet mandatory as 1.0.0 has not been
+  officially released. However, we plan to enforce its use as soon as
+  it is the case -- assuming that it has all the features needed for
+  building MirageOS applications.
+
+- Use proper names: same name for opam and ocamlfind package, and also ideally
+  for the top-level module.
+
+- All direct dependencies are specified in the opam file (e.g. sometimes some
+  direct dependencies are included by transitivity -- this is not very stable
+  and could break the package in the future if one of a transitive dependency
+  is updated) (NOTE: should probably be added in the packaging guideline)
+
+### ![Level 2](https://img.shields.io/badge/level-2-blue.svg) Define Package Scope
+
+- Have well-identified maintainers. GitHub recently
+  [introduced code owners](https://github.com/blog/2392-introducing-code-owners),
+  which may be worth to use for some (of our core) libraries.
+
+- Have a design document specifying the scope of the library (and esp. what is
+  out of scope)
+
+- Use the correct set of tags in opam metadata
+
+- Have documentation
```

Comment:
Crucial is to define the behaviour on edge cases (see e.g. the discussion about mirage-console being blocking on Xen vs partial write on Solo5 https://github.com/mirage/mirage-flow/issues/4), and whether and how arguments are checked (e.g. values given to `set*` in cstruct are not checked, see https://github.com/mirage/ocaml-cstruct/issues/78)